File: webApp/src/api/projectDesign.ts
/**
 * UC2 project-design (Phase-2) API service — one typed function per server op over
 * the project-scoped nested routes (server/internal/client/web/projectdesign.go +
 * handlers.go). projectId is always a PATH param. The Phase-2 TWIN of
 * api/systemDesign.ts. No React / caching here — TanStack Query owns that in the
 * hooks layer.
 *
 * The generated schema leaves `ProjectSessionStateResponse.view` opaque
 * (Record<string, never>) — the Phase-2 model JSON is fully typed on the wire but
 * not enumerated in the OpenAPI spec. getProjectSessionState therefore decodes the
 * opaque view into the hand-mirrored ProjectSessionState (api/types.ts).
 */
import { apiClient, toApiError } from './client';
import type {
  ProjectArtifactKind,
  ProjectSessionState,
  ProjectSessionStateView,
  ProjectPhaseAdvanceResponse,
  ReviewDecision,
  SDPDecision,
} from './types';

/**
 * POST /project-design/artifacts/draft — request/continue a Phase-2 co-author gate
 * for one of the eight DRAFTABLE kinds (NOT sdpReview). Feedback is woven into the
 * next draft on a re-entry. 202 Accepted with the continuity sessionRef.
 */
export async function requestProjectArtifactDraft(
  projectId: string,
  artifactKind: ProjectArtifactKind,
  feedback?: string
): Promise<string> {
  const { data, error, response } = await apiClient.POST(
    '/api/v1/projects/{projectId}/project-design/artifacts/draft',
    {
      params: { path: { projectId } },
      body: { artifactKind, ...(feedback !== undefined ? { feedback } : {}) },
    }
  );
  if (error !== undefined) throw toApiError(response.status, error);
  return data.sessionRef;
}

/** POST /project-design/artifacts/review — deliver the architect's per-artifact gate decision. */
export async function submitProjectReviewDecision(
  projectId: string,
  artifactKind: ProjectArtifactKind,
  decision: ReviewDecision,
  feedback?: string
): Promise<void> {
  const { error, response } = await apiClient.POST(
    '/api/v1/projects/{projectId}/project-design/artifacts/review',
    {
      params: { path: { projectId } },
      body: { artifactKind, decision, ...(feedback !== undefined ? { feedback } : {}) },
    }
  );
  if (error !== undefined) throw toApiError(response.status, error);
}

/**
 * POST /project-design/sdp/assemble — assemble the SDP review (the UC2 spine
 * workflow that joins the three estimate Engines into the options table). 202
 * Accepted with the SDP-review sessionRef. Poll the `sdpReview` session afterwards.
 */
export async function requestSDPCommit(projectId: string): Promise<string> {
  const { data, error, response } = await apiClient.POST(
    '/api/v1/projects/{projectId}/project-design/sdp/assemble',
    { params: { path: { projectId } } }
  );
  if (error !== undefined) throw toApiError(response.status, error);
  return data.sessionRef;
}

/** Optional rationale woven into an SDP decision. */
export interface SDPDecisionDetail {
  /** Required by the Manager on "commit" — names the chosen option. */
  readonly optionId?: string;
  /** Required by the Manager on "rejectAll". */
  readonly feedback?: string;
}

/**
 * POST /project-design/sdp/decision — deliver the architect's option-commitment
 * decision. commit binds the named option + advances; rejectAll re-enters Phase 2
 * with feedback. 204 No Content once the signal is durably enqueued.
 */
export async function submitSDPDecision(
  projectId: string,
  decision: SDPDecision,
  detail: SDPDecisionDetail = {}
): Promise<void> {
  const { error, response } = await apiClient.POST(
    '/api/v1/projects/{projectId}/project-design/sdp/decision',
    {
      params: { path: { projectId } },
      body: {
        decision,
        ...(detail.optionId !== undefined ? { optionId: detail.optionId } : {}),
        ...(detail.feedback !== undefined ? { feedback: detail.feedback } : {}),
      },
    }
  );
  if (error !== undefined) throw toApiError(response.status, error);
}

/**
 * POST /project-design/advance — seal Phase 2 (gating workflow). A non-Advanced
 * result is the NORMAL "you still owe artifacts X, Y / no option bound" answer
 * (HTTP 200), not an error.
 */
export async function advanceToConstruction(
  projectId: string
): Promise<ProjectPhaseAdvanceResponse> {
  const { data, error, response } = await apiClient.POST(
    '/api/v1/projects/{projectId}/project-design/advance',
    { params: { path: { projectId } } }
  );
  if (error !== undefined) throw toApiError(response.status, error);
  return data;
}

/**
 * GET /project-design/sessions/{artifactKind} — poll one Phase-2 session's state.
 * artifactKind may be any Phase-2 kind, including "sdpReview". The opaque `view` is
 * decoded into the typed ProjectSessionStateView here (see api/types.ts).
 */
export async function getProjectSessionState(
  projectId: string,
  artifactKind: ProjectArtifactKind
): Promise<ProjectSessionState> {
  const { data, error, response } = await apiClient.GET(
    '/api/v1/projects/{projectId}/project-design/sessions/{artifactKind}',
    { params: { path: { projectId, artifactKind } } }
  );
  if (error !== undefined) throw toApiError(response.status, error);
  // The generated schema types `view` opaquely (Record<string, never>); the Go
  // wire form is fully typed. Narrow through unknown to the hand-mirrored view.
  return {
    projectId: data.projectId,
    artifactKind: data.artifactKind,
    stage: data.stage,
    view: data.view as unknown as ProjectSessionStateView,
  };
}

File: server/internal/client/web/projectdesign.go
package web

import (
	"fmt"
	"net/http"

	"github.com/mixofreality-studio/archistrator/server/internal/manager/projectdesign"
	"github.com/mixofreality-studio/archistrator/server/internal/resourceaccess/projectstate"
)

// This file is the HTTP binding for the UC2 driveDesignPhase (Phase-2) facet. Each
// handler is a THIN routing facet (webClient.md §0): decode HTTP/JSON → resolve
// principal + authorize (security Utility) → translate to the Manager's typed input
// → invoke EXACTLY ONE projectDesignManager op → shape the reply. No business
// logic, no state, no cross-Manager sequencing. Mirrors handlers.go (Phase 1).

// --- request DTOs ----------------------------------------------------------

// requestProjectArtifactDraftRequest starts/continues a Phase-2 artifact
// co-authoring gate. artifactKind is one of the eight DRAFTABLE Phase-2 kinds
// (NOT sdpReview — that is assembled via /sdp/assemble).
type requestProjectArtifactDraftRequest struct {
	ArtifactKind string  `json:"artifactKind"`
	Feedback     *string `json:"feedback,omitempty"`
}

// submitProjectReviewDecisionRequest delivers the architect's per-artifact gate
// decision for a Phase-2 artifact (Signal reviewDecision, OQ-3). The project is the
// {projectId} path segment, not a body field.
type submitProjectReviewDecisionRequest struct {
	ArtifactKind string  `json:"artifactKind"`
	Decision     string  `json:"decision"` // "approve" | "reject" | "withdraw"
	Feedback     *string `json:"feedback,omitempty"`
}

// submitSDPDecisionRequest delivers the architect's SDP decision (Signal
// sdpDecision). optionId is required on commit; feedback is required on rejectAll.
// The project is the {projectId} path segment, not a body field.
type submitSDPDecisionRequest struct {
	Decision string  `json:"decision"` // "commit" | "rejectAll"
	OptionID *string `json:"optionId,omitempty"`
	Feedback *string `json:"feedback,omitempty"`
}

// projectSessionStateResponse is the read-only technical view of one Phase-2
// session (getSessionState Query).
type projectSessionStateResponse struct {
	ProjectID    string                         `json:"projectId"`
	ArtifactKind string                         `json:"artifactKind"`
	Stage        string                         `json:"stage"`
	View         projectdesign.SessionStateView `json:"view"`
}

// --- handlers --------------------------------------------------------------

// handleRequestProjectArtifactDraft — driveDesignPhase{Phase2 RequestArtifactDraft}.
// Routes to projectDesignManager.RequestArtifactDraft (Workflow).
func (c *Client) handleRequestProjectArtifactDraft(w http.ResponseWriter, r *http.Request) {
	projectID, err := parseProjectID(r.PathValue("projectId"))
	if err != nil {
		writeClientError(w, err)
		return
	}
	var req requestProjectArtifactDraftRequest
	if !decodeJSON(w, r, &req) {
		return
	}
	kind, err := parseProjectDesignKind(req.ArtifactKind, false)
	if err != nil {
		writeClientError(w, err)
		return
	}
	if !c.authorizeProject(w, r, "drive-phase", string(projectID)) {
		return
	}
	var feedback *projectdesign.ReviewFeedback
	if req.Feedback != nil {
		feedback = &projectdesign.ReviewFeedback{Notes: *req.Feedback}
	}
	ref, err := c.projectDesign.RequestArtifactDraft(r.Context(), projectdesign.ProjectID(projectID), kind, feedback)
	if err != nil {
		writeManagerError(w, err)
		return
	}
	writeJSON(w, http.StatusAccepted, sessionRefResponse{SessionRef: ref.String()})
}

// handleSubmitProjectReviewDecision — driveDesignPhase{Phase2 SubmitReviewDecision}.
// Routes to projectDesignManager.SubmitReviewDecision (per-artifact Signal).
func (c *Client) handleSubmitProjectReviewDecision(w http.ResponseWriter, r *http.Request) {
	projectID, err := parseProjectID(r.PathValue("projectId"))
	if err != nil {
		writeClientError(w, err)
		return
	}
	var req submitProjectReviewDecisionRequest
	if !decodeJSON(w, r, &req) {
		return
	}
	kind, err := parseProjectDesignKind(req.ArtifactKind, false)
	if err != nil {
		writeClientError(w, err)
		return
	}
	decision, err := parseProjectReviewDecision(req.Decision)
	if err != nil {
		writeClientError(w, err)
		return
	}
	if !c.authorizeProject(w, r, "approve-artifact", string(projectID)) {
		return
	}
	var feedback *projectdesign.ReviewFeedback
	if req.Feedback != nil {
		feedback = &projectdesign.ReviewFeedback{Notes: *req.Feedback}
	}
	if err := c.projectDesign.SubmitReviewDecision(r.Context(), projectdesign.ProjectID(projectID), kind, decision, feedback); err != nil {
		writeManagerError(w, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// handleRequestSDPCommit — driveDesignPhase{RequestSDPCommit}. Routes to
// projectDesignManager.RequestSDPCommit (Workflow).
func (c *Client) handleRequestSDPCommit(w http.ResponseWriter, r *http.Request) {
	projectID, err := parseProjectID(r.PathValue("projectId"))
	if err != nil {
		writeClientError(w, err)
		return
	}
	if !c.authorizeProject(w, r, "drive-phase", string(projectID)) {
		return
	}
	ref, err := c.projectDesign.RequestSDPCommit(r.Context(), projectdesign.ProjectID(projectID))
	if err != nil {
		writeManagerError(w, err)
		return
	}
	writeJSON(w, http.StatusAccepted, sessionRefResponse{SessionRef: ref.String()})
}

// handleSubmitSDPDecision — driveDesignPhase{SubmitSDPDecision}. Routes to
// projectDesignManager.SubmitSDPDecision (Signal). Commit binds the named option;
// RejectAll re-enters Phase 2 with feedback.
func (c *Client) handleSubmitSDPDecision(w http.ResponseWriter, r *http.Request) {
	projectID, err := parseProjectID(r.PathValue("projectId"))
	if err != nil {
		writeClientError(w, err)
		return
	}
	var req submitSDPDecisionRequest
	if !decodeJSON(w, r, &req) {
		return
	}
	decision, err := parseSDPDecision(req.Decision)
	if err != nil {
		writeClientError(w, err)
		return
	}
	if !c.authorizeProject(w, r, "approve-artifact", string(projectID)) {
		return
	}
	var optionID *projectdesign.OptionID
	if req.OptionID != nil {
		oid := projectdesign.OptionID(*req.OptionID)
		optionID = &oid
	}
	var feedback *projectdesign.ReviewFeedback
	if req.Feedback != nil {
		feedback = &projectdesign.ReviewFeedback{Notes: *req.Feedback}
	}
	if err := c.projectDesign.SubmitSDPDecision(r.Context(), projectdesign.ProjectID(projectID), decision, optionID, feedback); err != nil {
		writeManagerError(w, err)
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// handleAdvanceToConstruction — driveDesignPhase{AdvanceToConstruction}. Routes to
// projectDesignManager.AdvanceToConstruction (Workflow). A non-Advanced result is
// the NORMAL gating answer (HTTP 200), not an error.
func (c *Client) handleAdvanceToConstruction(w http.ResponseWriter, r *http.Request) {
	projectID, err := parseProjectID(r.PathValue("projectId"))
	if err != nil {
		writeClientError(w, err)
		return
	}
	if !c.authorizeProject(w, r, "drive-phase", string(projectID)) {
		return
	}
	result, err := c.projectDesign.AdvanceToConstruction(r.Context(), projectdesign.ProjectID(projectID))
	if err != nil {
		writeManagerError(w, err)
		return
	}
	missing := make([]string, 0, len(result.MissingArtifacts))
	for _, k := range result.MissingArtifacts {
		missing = append(missing, projectDesignKindName(k))
	}
	writeJSON(w, http.StatusOK, phaseAdvanceResponse{Advanced: result.Advanced, MissingArtifacts: missing})
}

// handleGetProjectSessionState — Phase-2 polling relayed to
// projectDesignManager.GetSessionState (Query). The artifactKind path segment may
// be any Phase-2 kind, including "sdpReview" (the SDP-review session).
func (c *Client) handleGetProjectSessionState(w http.ResponseWriter, r *http.Request) {
	projectID, err := parseProjectID(r.PathValue("projectId"))
	if err != nil {
		writeClientError(w, err)
		return
	}
	kind, err := parseProjectDesignKind(r.PathValue("artifactKind"), true)
	if err != nil {
		writeClientError(w, err)
		return
	}
	if !c.authorizeProject(w, r, "drive-phase", string(projectID)) {
		return
	}
	view, err := c.projectDesign.GetSessionState(r.Context(), projectdesign.ProjectID(projectID), kind)
	if err != nil {
		writeManagerError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, projectSessionStateResponse{
		ProjectID:    string(projectID),
		ArtifactKind: projectDesignKindName(kind),
		Stage:        projectSessionStageName(view.Stage),
		View:         view,
	})
}

// --- transport mappings ----------------------------------------------------

// phase2KindByName maps the wire artifactKind string to the typed Phase-2
// ArtifactKind. The nine Phase-2 kinds are accepted on this surface; an unknown or
// non-Phase-2 name is a client error.
var phase2KindByName = map[string]projectdesign.ArtifactKind{
	"planningAssumptions":  projectstate.KindPlanningAssumptions,
	"activityList":         projectstate.KindActivityList,
	"network":              projectstate.KindNetwork,
	"normalSolution":       projectstate.KindNormalSolution,
	"decompressedSolution": projectstate.KindDecompressedSolution,
	"subcriticalSolution":  projectstate.KindSubcriticalSolution,
	"compressedSolution":   projectstate.KindCompressedSolution,
	"riskModel":            projectstate.KindRiskModel,
	"sdpReview":            projectstate.KindSdpReview,
}

// projectDesignKindWireName is the inverse of phase2KindByName for response shaping.
var projectDesignKindWireName = func() map[projectdesign.ArtifactKind]string {
	m := make(map[projectdesign.ArtifactKind]string, len(phase2KindByName))
	for name, kind := range phase2KindByName {
		m[kind] = name
	}
	return m
}()

// parseProjectDesignKind maps a wire artifactKind to the typed Phase-2 kind. When
// allowSDPReview is false the "sdpReview" kind is rejected (it is assembled via
// /sdp/assemble, not co-authored as a draftable artifact).
func parseProjectDesignKind(s string, allowSDPReview bool) (projectdesign.ArtifactKind, error) {
	kind, ok := phase2KindByName[s]
	if !ok {
		return 0, fmt.Errorf("artifactKind %q is not a recognized Phase-2 kind", s)
	}
	if !allowSDPReview && kind == projectstate.KindSdpReview {
		return 0, fmt.Errorf("artifactKind %q is assembled via /sdp/assemble, not co-authored as a draft", s)
	}
	return kind, nil
}

// projectDesignKindName renders a Phase-2 ArtifactKind onto the wire.
func projectDesignKindName(kind projectdesign.ArtifactKind) string {
	if name, ok := projectDesignKindWireName[kind]; ok {
		return name
	}
	return kind.String()
}

// parseProjectReviewDecision maps the wire decision string to the typed Phase-2
// ReviewDecision (the per-artifact co-authoring gate).
func parseProjectReviewDecision(s string) (projectdesign.ReviewDecision, error) {
	switch s {
	case "approve":
		return projectdesign.ReviewApprove, nil
	case "reject":
		return projectdesign.ReviewReject, nil
	case "withdraw":
		return projectdesign.ReviewWithdraw, nil
	default:
		return projectdesign.ReviewDecisionUnknown, fmt.Errorf("decision %q must be one of approve|reject|withdraw", s)
	}
}

// parseSDPDecision maps the wire decision string to the typed SDPDecision (the
// option-commitment gate).
func parseSDPDecision(s string) (projectdesign.SDPDecision, error) {
	switch s {
	case "commit":
		return projectdesign.SDPCommit, nil
	case "rejectAll":
		return projectdesign.SDPRejectAll, nil
	default:
		return projectdesign.SDPDecisionUnknown, fmt.Errorf("decision %q must be one of commit|rejectAll", s)
	}
}

// projectSessionStageName renders the technical Phase-2 SessionStage onto the wire.
func projectSessionStageName(stage projectdesign.SessionStage) string {
	switch stage {
	case projectdesign.StageDrafting:
		return "drafting"
	case projectdesign.StageAssemblingSDP:
		return "assemblingSdp"
	case projectdesign.StageAwaitingReview:
		return "awaitingReview"
	case projectdesign.StageRedrafting:
		return "redrafting"
	case projectdesign.StageCommitted:
		return "committed"
	case projectdesign.StageWithdrawn:
		return "withdrawn"
	case projectdesign.StageRefused:
		return "refused"
	case projectdesign.StageDraftFailed:
		// The human-visible, human-actionable terminal-failure stage (the Phase-2 twin
		// of systemdesign's StageDraftFailed): the dispatched agentic Phase-2 design job
		// reached a TYPED terminal failure phase. Surfaced as a DISTINCT wire stage (NOT
		// "unknown") so the SPA shows a retry/withdraw affordance, never a perpetual
		// "drafting" spinner (D-MPD-Δ §3.4, the anti-wedge requirement).
		return "draftFailed"
	default:
		return "unknown"
	}
}
